nipype/algorithms/compcor.py

```diff
@@ -0,0 +1,115 @@
+from ..interfaces.base import (BaseInterfaceInputSpec, TraitedSpec,
+                               BaseInterface, traits, File)
+
+from nipype.pipeline import engine as pe
+from nipype.interfaces.utility import IdentityInterface
+
+import nibabel as nb
+import numpy as np
+from scipy import linalg, stats
+import os
+
+class CompCorInputSpec(BaseInterfaceInputSpec):
+    realigned_file = File(exists=True, mandatory=True,
+                          desc='already realigned brain image (4D)')
+    mask_file = File(exists=True, mandatory=False,
+                     desc='mask file that determines ROI (3D)')
+    components_file = File('components_file.txt', exists=False, mandatory=False, usedefault=True,
+                           desc='filename to store physiological components in')
+    num_components = traits.Int(6, usedefault=True) # 6 for BOLD, 4 for ASL
+
+class CompCorOutputSpec(TraitedSpec):
+    components_file = File(desc='text file containing the noise components', exists=True)
+
+class CompCor(BaseInterface):
+    '''
+    Interface with core CompCor computation, used in aCompCor and tCompCor
+
+    Example
+    -------
+
+    >>> ccinterface = CompCor()
+    >>> ccinterface.inputs.realigned_file = 'nipype/testing/data/functional.nii'
+    >>> ccinterface.inputs.mask_file = 'nipype/testing/data/mask.nii'
+    >>> ccinterface.inputs.num_components = 1
+    '''
+    input_spec = CompCorInputSpec
+    output_spec = CompCorOutputSpec
+
+    def _run_interface(self, runtime):
+        imgseries = nb.load(self.inputs.realigned_file).get_data()
+        mask = nb.load(self.inputs.mask_file).get_data()
+        voxel_timecourses = imgseries[mask > 0]
+        # Zero-out any bad values
+        voxel_timecourses[np.isnan(np.sum(voxel_timecourses, axis=1)), :] = 0
+
+        # from paper:
+        # "Voxel time series from the noise ROI (either anatomical or tSTD) were
+        # placed in a matrix M of size Nxm, with time along the row dimension
+        # and voxels along the column dimension."
+        # voxel_timecourses.shape == [nvoxels, time]
+        M = voxel_timecourses.T
+        numvols = M.shape[0]
+        numvoxels = M.shape[1]
+
+        # "The constant and linear trends of the columns in the matrix M were removed ..."
+        timesteps = range(numvols)
```

Comment:
you may want to use legendre polynomials for this as shown here:

https://github.com/nipy/nipype/blob/master/nipype/algorithms/misc.py#L310

also this should become a boolean flag + order being another parameter that's set to some default value but user changeable.